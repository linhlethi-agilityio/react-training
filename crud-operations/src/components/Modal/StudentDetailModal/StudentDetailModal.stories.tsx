import type { Meta, StoryObj } from '@storybook/react';

// Components
import StudentDetailModal from '.';

const meta: Meta<typeof StudentDetailModal> = {
  component: StudentDetailModal,
};

export default meta;
type Story = StoryObj<typeof StudentDetailModal>;

export const Default: Story = {
  args: {
    isOpen: true,
    onClose: () => null,
    previewData: null,
  },
};
