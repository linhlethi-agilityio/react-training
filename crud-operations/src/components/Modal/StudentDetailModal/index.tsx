import { memo, useMemo } from 'react';
import { Controller, useForm } from 'react-hook-form';
import { Box, Button, FormControl, FormHelperText, FormLabel, Input } from '@chakra-ui/react';

// Constants
import { ERROR_MESSAGES } from '@constants';

// Utils
import { clearErrorOnChange, formatDate, isValidEmail } from '@utils';

// Types
import { Student } from '@types';

// Hooks
import { useStudents, useToastCustom } from '@hooks';

// Components
import { CustomModal } from '@components';
interface StudentModalDetailModalProps {
  isOpen: boolean;
  previewData?: Student | null;
  onClose: () => void;
}

type StudentFormState = Omit<Student, 'id'>;

const initFormData = {
  name: '',
  email: '',
  phone: '',
  enrollNumber: '',
  dateOfAdmission: '',
};

const StudentDetailModal = ({ isOpen, previewData, onClose }: StudentModalDetailModalProps) => {
  const { createStudent, updateStudent } = useStudents();
  const toast = useToastCustom();

  const {
    control,
    clearErrors,
    handleSubmit,
    formState: { errors },
  } = useForm<StudentFormState>({
    mode: 'onBlur',
    reValidateMode: 'onBlur',
    defaultValues: previewData ? previewData : initFormData,
  });

  const isEditMode = useMemo(() => !!previewData?.id, [previewData?.id]);

  const handleOnSubmit = async (fromData: StudentFormState) => {
    let response = null;
    const dateOfAdmission = new Date(fromData.dateOfAdmission).getTime();

    if (isEditMode) {
      response = await updateStudent({
        ...previewData,
        ...fromData,
        dateOfAdmission,
      });
    } else {
      response = await createStudent({
        ...fromData,
        dateOfAdmission,
      });
    }

    onClose();

    if (response) {
      toast({
        title: 'Request sent successfully!',
        description: `A new student has been ${isEditMode ? 'edited' : 'added'}!`,
        status: 'success',
      });
    } else {
      toast({
        title: 'Request failed!',
        description: 'Something went wrong, please try again!',
        status: 'error',
      });
    }
  };

  return (
    <CustomModal isOpen={isOpen} onClose={onClose} title={isEditMode ? 'Update Student' : 'Add Student'}>
      <FormControl textAlign="center">
        {/* Name */}
        <Controller
          name="name"
          control={control}
          rules={{ required: ERROR_MESSAGES.FIELD_REQUIRED }}
          render={({ field: { name, onChange, ...rest }, fieldState: { error } }) => (
            <Box marginBottom={error?.message ? 0 : 27}>
              <FormLabel fontSize="sm" lineHeight="sm" color="text.default">
                Name
              </FormLabel>
              <Input
                isInvalid={!!error?.message}
                onChange={(e) => {
                  onChange(e.target.value);

                  // Clear error message on change
                  clearErrorOnChange(name, errors, clearErrors);
                }}
                {...rest}
              />
              {error?.message && (
                <FormHelperText color="error" fontSize="xs" mb={1} textAlign="left">
                  {error.message}
                </FormHelperText>
              )}
            </Box>
          )}
        />

        {/* Email */}
        <Controller
          name="email"
          control={control}
          rules={{
            required: ERROR_MESSAGES.FIELD_REQUIRED,
            validate: (email) => isValidEmail(email) || ERROR_MESSAGES.INVALID_EMAIL,
          }}
          render={({ field: { name, onChange, ...rest }, fieldState: { error } }) => (
            <Box marginBottom={error?.message ? 0 : 27}>
              <FormLabel fontSize="sm" lineHeight="sm" color="text.default">
                Email
              </FormLabel>
              <Input
                aria-label="Email"
                isInvalid={!!error?.message}
                onChange={(e) => {
                  onChange(e.target.value);

                  // Clear error message on change
                  clearErrorOnChange(name, errors, clearErrors);
                }}
                {...rest}
              />
              {error?.message && (
                <FormHelperText color="error" fontSize="xs" mb={1} textAlign="left">
                  {error.message}
                </FormHelperText>
              )}
            </Box>
          )}
        />

        {/* Phone */}
        <Controller
          name="phone"
          control={control}
          rules={{ required: ERROR_MESSAGES.FIELD_REQUIRED }}
          render={({ field: { name, onChange, ...rest }, fieldState: { error } }) => (
            <Box marginBottom={error?.message ? 0 : 27}>
              <FormLabel fontSize="sm" lineHeight="sm" color="text.default">
                Phone
              </FormLabel>
              <Input
                isInvalid={!!error?.message}
                onChange={(e) => {
                  onChange(e.target.value);

                  // Clear error message on change
                  clearErrorOnChange(name, errors, clearErrors);
                }}
                {...rest}
              />
              {error?.message && (
                <FormHelperText color="error" fontSize="xs" mb={1} textAlign="left">
                  {error.message}
                </FormHelperText>
              )}
            </Box>
          )}
        />

        {/* Enroll number */}
        <Controller
          name="enrollNumber"
          control={control}
          rules={{ required: ERROR_MESSAGES.FIELD_REQUIRED }}
          render={({ field: { name, onChange, ...rest }, fieldState: { error } }) => (
            <Box marginBottom={error?.message ? 0 : '25px'}>
              <FormLabel fontSize="sm" lineHeight="sm" color="text.default">
                Enroll number
              </FormLabel>
              <Input
                isInvalid={!!error?.message}
                onChange={(e) => {
                  onChange(e.target.value);

                  // Clear error message on change
                  clearErrorOnChange(name, errors, clearErrors);
                }}
                {...rest}
              />
              {error?.message && (
                <FormHelperText color="error" fontSize="xs" mb={1} textAlign="left">
                  {error.message}
                </FormHelperText>
              )}
            </Box>
          )}
        />

        {/* Date of admission */}
        <Controller
          name="dateOfAdmission"
          control={control}
          rules={{ required: ERROR_MESSAGES.FIELD_REQUIRED }}
          render={({ field: { name, value, onChange, ...rest }, fieldState: { error } }) => (
            <Box marginBottom={error?.message ? 0 : 25}>
              <FormLabel fontSize="sm" lineHeight="sm" color="text.default">
                Date of admission
              </FormLabel>
              <Input
                type="date"
                value={formatDate(value, true)}
                isInvalid={!!error?.message}
                onChange={(e) => {
                  onChange(e.target.value);

                  // Clear error message on change
                  clearErrorOnChange(name, errors, clearErrors);
                }}
                {...rest}
              />
              {error?.message && (
                <FormHelperText color="error" fontSize="xs" mb={1} textAlign="left">
                  {error.message}
                </FormHelperText>
              )}
            </Box>
          )}
        />
        <Button type="submit" w="full" marginBottom={5} onClick={handleSubmit(handleOnSubmit)}>
          Submit
        </Button>
      </FormControl>
    </CustomModal>
  );
};

export default memo(StudentDetailModal);
