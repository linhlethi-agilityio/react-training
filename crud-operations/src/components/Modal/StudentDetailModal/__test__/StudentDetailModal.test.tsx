import { act, fireEvent, render, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';

// Components
import StudentDetailModal from '..';

const mockCreateStudent = jest.fn();
const mockUpdateStudent = jest.fn();
const mockOnClose = jest.fn();

jest.mock('@hooks', () => ({
  useStudents: jest.fn(() => ({
    createStudent: mockCreateStudent,
    updateStudent: mockUpdateStudent,
  })),
  useToastCustom: () => jest.fn(),
}));

jest.mock('jose', () => ({}));

jest.mock('@constants', () => ({
  ERROR_MESSAGES: {
    FIELD_REQUIRED: 'Mocked field required message',
  },
  REGEX_PATTERN: {
    EMAIL:
      /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,
    NUMBER: /^[0-9]{10}$/,
    NAME: /^[a-zA-Z ]{1,40}$/,
  },
}));

const mockProps = {
  isOpen: true,
  onClose: mockOnClose,
  previewData: null,
};

const student = {
  id: '123',
  name: 'joun',
  email: 'admin@gmail.com',
  phone: '6565654656',
  enrollNumber: '456565656',
  dateOfAdmission: '2024-05-05',
};

describe('StudentDetailModal component', () => {
  test('should match snapshot for StudentDetailModal', () => {
    const { container } = render(<StudentDetailModal {...mockProps} />);

    expect(container).toMatchSnapshot();
  });

  test('create student', async () => {
    const { getByRole, getByTestId } = render(<StudentDetailModal {...mockProps} />);

    act(() => {
      fireEvent.change(getByTestId('name'), { target: { value: 'jonh' } });
      fireEvent.change(getByTestId('email'), { target: { value: 'test@gmail.com' } });
      fireEvent.change(getByTestId('phone'), { target: { value: '5765756756' } });
      fireEvent.change(getByTestId('dateOfAdmission'), { target: { value: '2024-03-21' } });

      const button = getByRole('button', { name: 'button-submit' });

      fireEvent.click(button);

      waitFor(() => {
        expect(mockCreateStudent).toHaveBeenCalled();
      });
    });
  });

  test('update student', async () => {
    const { getByRole, getByTestId } = render(<StudentDetailModal previewData={student} onClose={jest.fn()} isOpen />);

    act(() => {
      fireEvent.change(getByTestId('name'), { target: { value: 'jonh' } });
      fireEvent.change(getByTestId('email'), { target: { value: 'test@gmail.com' } });
      fireEvent.change(getByTestId('phone'), { target: { value: '5765756756' } });
      fireEvent.change(getByTestId('enrollNumber'), { target: { value: 5765756756 } });
      fireEvent.change(getByTestId('dateOfAdmission'), { target: { value: '2024-03-21' } });

      const button = getByRole('button', { name: 'button-submit' });

      fireEvent.click(button);

      waitFor(() => {
        expect(mockUpdateStudent).toHaveBeenCalled();
      });
    });
  });
});
