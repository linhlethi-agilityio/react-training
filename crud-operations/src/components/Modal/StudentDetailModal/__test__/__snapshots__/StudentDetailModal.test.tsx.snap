// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudentDetailModal component should match snapshot for StudentDetailModal 1`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
/>
`;
